ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/FastSyncDownloader.java

```diff
@@ -80,17 +87,31 @@ public FastSyncDownloader(
             .thenApply(this::updateMaxTrailingPeers)
             .thenApply(this::storeState)
             .thenCompose(this::downloadChainAndWorldState),
-        this::handleWorldStateUnavailable);
+        this::handleFailure);
   }
 
-  private CompletableFuture<FastSyncState> handleWorldStateUnavailable(final Throwable error) {
+  private CompletableFuture<FastSyncState> handleFailure(final Throwable error) {
     trailingPeerRequirements = Optional.empty();
-    if (ExceptionUtils.rootCause(error) instanceof StalledDownloadException) {
+    if (ExceptionUtils.rootCause(error) instanceof FastSyncException) {
+      return completedExceptionally(error);
+    } else if (ExceptionUtils.rootCause(error) instanceof StalledDownloadException) {
       LOG.warn(
           "Fast sync was unable to download the world state. Retrying with a new pivot block.");
       return start(FastSyncState.EMPTY_SYNC_STATE);
     } else {
-      return completedExceptionally(error);
+      LOG.error("Encountered an unexpected error during fast sync.  Trying to restart fast sync.");
+      final ScheduledFuture<CompletableFuture<FastSyncState>> fastSyncCountdown =
+          Executors.newSingleThreadScheduledExecutor()
```

Comment:
We already have some utilities around executing timeouts.  I suggest hooking into those by adding a helper method to `FastSyncActions` which has access to `EthScheduler` - `FastSyncActions.scheduleTask(Supplier<CompletableFuture<T>> future, Duration duration)` perhaps?  See: https://github.com/hyperledger/besu/blob/28205875c5683cce832450f05898ee5316139176/ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthScheduler.java#L170

Reply:
good idea. done